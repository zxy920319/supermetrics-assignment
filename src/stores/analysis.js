import { defineStore } from "pinia"
import { getAllPosts } from "@/client/api"

export const useAnalysisStore = defineStore("analysis", {
  state: () => {
    return { posts: null }
  },
  actions: {
    async fetchAllPosts(params) {
      if (!this.posts) this.posts = await getAllPosts(params)
    },
  },
})
