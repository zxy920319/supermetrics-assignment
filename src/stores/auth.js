import { defineStore } from "pinia"
import { getToken } from "@/client"

export const useAuthStore = defineStore("auth", {
  state: () => {
    return { token: null }
  },
  actions: {
    async refreshToken() {
      this.token = await getToken()
    },
  },
})
