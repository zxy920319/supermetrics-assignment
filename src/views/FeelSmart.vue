<template>
  <DashboardContent>
    <div>
      <div v-show="loading" class="absolute bg-gray-400 bg-opacity-50">
        <ASpin :loading="loading"> </ASpin>
      </div>
      <div class="flex flex-wrap justify-start gap-6">
        <ACard v-for="user in usersInfo" :key="user" :title="user.name" hoverable class="user-card">
          <div>Total Number of Posts: {{ user.total }}</div>
          <div>Average number of characters of his/her posts: {{ user.averageLength }}</div>
          <div>
            The number of posts made every month:
            <ATag v-for="month in user.byMonth.keys()" :key="month">{{ `${month}: ${user.byMonth.get(month)} Posts` }}</ATag>
          </div>
          <div>Lonest Post: {{ user.longestPost.message }}</div>
        </ACard>
      </div>
    </div>
  </DashboardContent>
</template>

<script setup>
import { useAuthStore } from "@/stores/auth"
import { getAllPosts } from "@/client"
import DashboardContent from "@/components/DashboardContent.vue"
import { computed, onMounted, ref } from "@vue/runtime-core"
import dayjs from "dayjs"
import { Skeleton as ASpin, Card as ACard, Tag as ATag } from "ant-design-vue"
import "ant-design-vue/es/spin/style/css"
import "ant-design-vue/es/card/style/css"
import "ant-design-vue/es/tag/style/css"

const store = useAuthStore()
const posts = ref([])
const loading = ref(true)
const months = dayjs.months()

const users = computed(() => {
  const usersMap = new Map()
  posts.value.forEach((post) => {
    if (usersMap.has(post.from_id)) {
      const originPosts = usersMap.get(post.from_id)
      usersMap.set(post.from_id, originPosts.concat(post))
    } else {
      usersMap.set(post.from_id, [post])
    }
  })
  return usersMap
})

const calcAverageLength = (userPosts) => {
  return Math.round(
    userPosts.reduce((totalChar, currentPost) => {
      return totalChar + currentPost.message.length
    }, 0) / userPosts.length
  )
}

const calcMonthPosts = (userPosts) => {
  const byMonth = new Map()
  userPosts.forEach((post) => {
    const month = months[dayjs(post.created_time).month()]
    if (byMonth.has(month)) {
      const postCount = byMonth.get(month)
      byMonth.set(month, postCount + 1)
    } else {
      byMonth.set(month, 1)
    }
  })
  return byMonth
}

const calcLongestPost = (userPosts) => {
  return userPosts.reduce(
    (longestPost, currentPost) => {
      return longestPost.message.length < currentPost.message.length ? currentPost : longestPost
    },
    { message: "" }
  )
}

const usersInfo = computed(() => {
  const result = []
  if (!(users.value instanceof Map)) return []

  for (const [key, value] of users.value.entries()) {
    result.push({
      key,
      name: value[0].from_name,
      total: value.length,
      averageLength: calcAverageLength(value),
      byMonth: calcMonthPosts(value),
      longestPost: calcLongestPost(value),
    })
  }

  return result
})

onMounted(async () => {
  const params = {
    sl_token: store.token,
  }
  posts.value = await getAllPosts(params)
  loading.value = false
})
</script>

<style lang="scss" scoped>
.user-card {
  @apply w-64;
}
</style>
