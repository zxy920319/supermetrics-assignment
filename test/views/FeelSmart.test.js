import FeelSmart from "@/views/FeelSmart.vue"
import { createTestingPinia } from "@pinia/testing"
import { config, flushPromises, shallowMount } from "@vue/test-utils"
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest"
import { useAnalysisStore } from "@/stores/analysis"

describe("FeelSmart", () => {
  const analysis = {
    posts: [
      {
        from_name: "John Doe",
        from_id: "user_3",
        message: "some message",
      },
      {
        from_name: "Leonarda Schult",
        from_id: "user_3",
        message:
          "gradient wrestle interest prescription hallway beautiful fireplace bell diameter abstract rally leader achievement porter trail gift core litigation executive horseshoe confront sow hike stride era radio generation resident virtue irony rape generate transport contraction sweet delicate band contrary balance publisher national marriage deserve",
      },
      {
        from_name: "Yolande Urrutia",
        from_id: "user_15",
        message:
          "process siege name mess situation porter suggest racism romantic unfair shell serve body visible slab disaster loose meadow turkey allocation favor food accident correspond counter virgin parachute spend situation restrict civilian proposal grandmother brake deposit good broken sunshine crusade hate bolt margin introduce memorandum printer correspond belief planet sigh far animal act bolt introduction diameter mayor value mastermind",
      },
      {
        from_name: "Lashanda Small",
        from_id: "user_12",
        message: "1",
      },
      {
        from_name: "Lashanda Small",
        from_id: "user_12",
        message: "2",
      },
      {
        from_name: "Ethelene Maggi",
        from_id: "user_18",
        message: "3",
      },
      {
        from_name: "Woodrow Lindholm",
        from_id: "user_14",
        message: "4",
      },
      {
        from_name: "Rosann Eide",
        from_id: "user_9",
        message: "5",
      },
      {
        from_name: "Ethelene Maggi",
        from_id: "user_18",
        message: "6",
      },
      {
        from_name: "Woodrow Lindholm",
        from_id: "user_14",
        message: "7",
      },
      {
        from_name: "Regenia Boice",
        from_id: "user_13",
        message: "8",
      },
    ],
    fetchAllPosts: vi.fn(),
  }
  const formatDateTime = () => ""
  const render = () => {
    return shallowMount(FeelSmart, {
      global: {
        plugins: [createTestingPinia({ initialState: { auth: { token: "token" }, analysis }, createSpy: vi.fn })],
        provide: { formatDateTime },
      },
    })
  }

  beforeAll(() => {
    config.renderStubDefaultSlot = true
  })

  afterAll(() => {
    config.renderStubDefaultSlot = false
  })

  it("calls fetchAllPosts on mounted", async () => {
    const wrapper = await render()
    await flushPromises()
    const analysisStore = useAnalysisStore()

    expect(wrapper.html()).toMatchSnapshot()
    expect(analysisStore.fetchAllPosts).toHaveBeenCalledWith({ sl_token: "token" })
    expect(wrapper.vm.loading).toBe(false)
  })

  it("generates users from posts", async () => {
    const wrapper = await render()
    await flushPromises()

    expect(wrapper.vm.users.size).toEqual(7)
    expect(wrapper.vm.users.get("user_12")).toEqual([
      expect.anything(),
      expect.objectContaining({
        from_name: "Lashanda Small",
        from_id: "user_12",
      }),
    ])
  })

  it("generates usersInfo from users", async () => {
    const wrapper = await render()
    await flushPromises()

    expect(wrapper.vm.usersInfo.length).toEqual(7)
    expect(wrapper.vm.usersInfo[0].name).toEqual("John Doe")
    expect(wrapper.vm.usersInfo[0].total).toEqual(2)
    expect(wrapper.vm.usersInfo[0].averageLength).toEqual(179)
    expect(wrapper.vm.usersInfo[0].byMonth.length).toEqual(1)
  })
})
