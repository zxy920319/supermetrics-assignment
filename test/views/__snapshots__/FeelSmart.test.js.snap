// Vitest Snapshot v1

exports[`FeelSmart > calls fetchAllPosts on mounted 1`] = `
"<dashboard-content-stub data-v-4c478579=\\"\\">
  <div data-v-4c478579=\\"\\">
    <div class=\\"absolute bg-gray-400 bg-opacity-50 w-full h-full top-0 left-0 z-10 flex items-center justify-center\\" data-v-4c478579=\\"\\" style=\\"display: none;\\">
      <a-spin-stub tip=\\"Loading...\\" spinning=\\"true\\" size=\\"default\\" wrapperclassname=\\"\\" loading=\\"false\\" data-v-4c478579=\\"\\"></a-spin-stub>
    </div>
    <div class=\\"flex flex-wrap justify-start gap-6\\" data-v-4c478579=\\"\\">
      <a-card-stub bordered=\\"true\\" loading=\\"false\\" hoverable=\\"true\\" class=\\"user-card\\" data-v-4c478579=\\"\\">
        <a-comment-stub class=\\"bg-gray-50\\" data-v-4c478579=\\"\\"></a-comment-stub>
      </a-card-stub>
      <a-card-stub bordered=\\"true\\" loading=\\"false\\" hoverable=\\"true\\" class=\\"user-card\\" data-v-4c478579=\\"\\">
        <a-comment-stub class=\\"bg-gray-50\\" data-v-4c478579=\\"\\"></a-comment-stub>
      </a-card-stub>
      <a-card-stub bordered=\\"true\\" loading=\\"false\\" hoverable=\\"true\\" class=\\"user-card\\" data-v-4c478579=\\"\\">
        <a-comment-stub class=\\"bg-gray-50\\" data-v-4c478579=\\"\\"></a-comment-stub>
      </a-card-stub>
      <a-card-stub bordered=\\"true\\" loading=\\"false\\" hoverable=\\"true\\" class=\\"user-card\\" data-v-4c478579=\\"\\">
        <a-comment-stub class=\\"bg-gray-50\\" data-v-4c478579=\\"\\"></a-comment-stub>
      </a-card-stub>
      <a-card-stub bordered=\\"true\\" loading=\\"false\\" hoverable=\\"true\\" class=\\"user-card\\" data-v-4c478579=\\"\\">
        <a-comment-stub class=\\"bg-gray-50\\" data-v-4c478579=\\"\\"></a-comment-stub>
      </a-card-stub>
      <a-card-stub bordered=\\"true\\" loading=\\"false\\" hoverable=\\"true\\" class=\\"user-card\\" data-v-4c478579=\\"\\">
        <a-comment-stub class=\\"bg-gray-50\\" data-v-4c478579=\\"\\"></a-comment-stub>
      </a-card-stub>
      <a-card-stub bordered=\\"true\\" loading=\\"false\\" hoverable=\\"true\\" class=\\"user-card\\" data-v-4c478579=\\"\\">
        <a-comment-stub class=\\"bg-gray-50\\" data-v-4c478579=\\"\\"></a-comment-stub>
      </a-card-stub>
    </div>
  </div>
</dashboard-content-stub>"
`;
